### ISSUE TITLE
Tiny table joining two large tables should pre join both

### ISSUE THREAD

**Issue by kaikalur** [prestodb/presto#23525]
We have seen a number of prod queries like:

select .. T1 JOIN T2 ON T1.x = T2.x JOIN T3 ON T2.y=T3.y

And T3 is tiny - a few thousand rows but T1 and T2 are several 100B rows. In that case, we should join the two large tables T1 and T2 separately with T3 and then join the result (maybe make it cost based)

**Comment by auden-woolfson** [comment]
In this scenario isn't it possible that T1 doesn't have field y, so we should actually just make sure we are joining T2 and T3 first, then joining with T1?

**Comment by kaikalur** [comment]
> In this scenario isn't it possible that T1 doesn't have field y, so we should actually just make sure we are joining T2 and T3 first, then joining with T1?

Yes that's also another thing we don't do. But I have seen the other situation as well where T3 joins T1 but we don't take advantage of that.

So 

T1 JOIN T2 JOIN T3 
when T3 don't join T1 -> T1 JOIN (T2 JOIN T3)
when T3 JOIN only T1 and not T2 -> T2 JOIN (T1 JOIN T3)
when it joins both-> (T1 JOIN T3) JOIN (T2 JOIN T3)
